### PR TITLE
enhancement(host_metrics source): integrate upstream container support

### PR DESCRIPTION
**Pull request by eeyun**
This one is a tough one to test. Theoretically this shouldn't alter any behavior whatsoever _currently_

In order to test the behavior we have to be running a build of the branch in a container with the appropriate environment variables set. I'll post some output from my local testing here and ultimately @MOZGIII is adding e2e tests for this in our kubernetes testing components.

### CODE REVIEW

**Comment by eeyun** [comment]
To test this and prove its working you can do the following:

Make a local build of the branch.
`make package-deb-x86_64`

In my case I created a `custom.toml` file and placed it in `target/artifacts`.
```
[sources.generator]
type = "host_metrics"
collectors = ["cpu", "disk", "filesystem", "load", "memory", "network"]
scrape_interval_secs = 5

[sinks.stdout]
type = "console"
inputs = ["generator"]
target = "stdout"
encoding.codec = "json"
```

Then I modified `distribution/docker/debian/Dockerfile` to specifically test this usecase.

```
FROM debian:buster-slim AS builder

COPY vector-*.deb ./

RUN dpkg -i vector-$(dpkg --print-architecture).deb

RUN rm /etc/vector/vector.toml

COPY custom.toml /etc/vector/vector.toml

FROM debian:buster-slim

RUN apt-get update && apt-get install -y ca-certificates tzdata systemd && rm -rf /var/lib/apt/lists/*

COPY --from=builder /usr/bin/vector /usr/bin/vector
COPY --from=builder /usr/share/doc/vector /usr/share/doc/vector
COPY --from=builder /etc/vector /etc/vector

RUN mkdir /host

VOLUME /var/lib/vector/

# Smoke test
RUN ["vector", "--version"]

ENTRYPOINT ["/usr/bin/vector"]
```
Built the container
```
❯❯❯ docker build --tag procroot target/artifacts -f distribution/docker/debian/Dockerfile
```

Finally, ran the container with environment variables set, networking configured to host and privileged mode:
```
❯❯❯ sudo docker run --privileged -e PROCFS_ROOT="/host/proc" -e SYSFS_ROOT="/host/sys" -v /proc:/host/proc:rw -v /sys:/host/sys:rw --network host --name mycontainer -it --rm procroot
```

Should net you some output that is distinct to the host:
```
{"name":"network_receive_packets_total","namespace":"host","timestamp":"2020-12-08T22:44:16.700524279Z","tags":{"collector":"network","device":"docker0","host":"curse"},"kind":"absolute","counter":{"value":136452.0}}
{"name":"network_transmit_bytes_total","namespace":"host","timestamp":"2020-12-08T22:44:16.700524279Z","tags":{"collector":"network","device":"docker0","host":"curse"},"kind":"absolute","counter":{"value":2203702506.0}}
{"name":"network_transmit_errs_total","namespace":"host","timestamp":"2020-12-08T22:44:16.700524279Z","tags":{"collector":"network","device":"docker0","host":"curse"},"kind":"absolute","counter":{"value":0.0}}
{"name":"network_transmit_packets_drop_total","namespace":"host","timestamp":"2020-12-08T22:44:16.700524279Z","tags":{"collector":"network","device":"docker0","host":"curse"},"kind":"absolute","counter":{"value":0.0}}
{"name":"network_transmit_packets_total","namespace":"host","timestamp":"2020-12-08T22:44:16.700524279Z","tags":{"collector":"network","device":"docker0","host":"curse"},"kind":"absolute","counter":{"value":221988.0}}
```

File: src/sources/host_metrics.rs
Comment (bruceg) [inline src/sources/host_metrics.rs:698]: I suggest something like `init_roots`, and call with `host_metrics::init_roots` in `app.rs`.

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs:705]: `set_procfs_root` takes a path value and configures heim to use it rather than the default procfs root path.

File: Cargo.toml
Comment (eeyun) [reply Cargo.toml:190]: `heim` doesn't include `heim_runtime` in it's root so it has to be depended upon separately.

File: src/sources/host_metrics.rs
Comment (bruceg) [reply src/sources/host_metrics.rs:705]: That's what I thought, but this code appears to pass the literal path `PROCFS_ROOT` instead of the value of that environment variable.

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs:707]: Oh yeah apologies this is an old commit! whoops!

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs]: Ah I see what happened - this was the wrong commit. Should be resolved!

File: src/sources/host_metrics.rs
Comment (bruceg) [inline src/sources/host_metrics.rs:698]: I wonder if the `#[cfg]` shouldn't be inside this function to reduce the complexity upstream.

File: src/sources/host_metrics.rs
Comment (bruceg) [inline src/sources/host_metrics.rs:710]: Is this feature useful for non-virtual environments, like for doing testing? That is, would there be any problems with removing the `virtualized` guard here?

File: src/sources/host_metrics.rs
Comment (bruceg) [inline src/sources/host_metrics.rs:699]: Should this be an `async fn` to just use `.await` here?

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs:710]: I can't imagine that it would be super useful? But, I could be wrong!

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs:699]: We _can_ make this async, but then any synchronous function that calls this one like we're doing in `app.rs` has an unused implementer of `futures::Future` which breaks our lints

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs:698]: In heim these functions can _only_ be implemented for linux so I put the guard here avoid this call site breaking at runtime on windows or darwin.

Comment (bruceg) [reply src/sources/host_metrics.rs:698]: Right. I was suggesting moving the `#[cfg]` inside the function, so that the function isn't omitted on non-Linux. As it stands now, it is always referenced in `fn prepare_from_opts`, so it won't compile on other OS.

File: src/sources/host_metrics.rs
Comment (bruceg) [reply src/sources/host_metrics.rs:699]: Right, we would have to move the blocking into the caller(s), which would require them to either become async themselves or move the `block_on` there.

Comment (eeyun) [reply src/sources/host_metrics.rs:699]: Ah indeed - I more meant we have to block on it somewhere, so if you'd prefer to do so at the call site just say the word.

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs:698]: Oh yes! My apologies, I completely misunderstood what you were saying 😅

File: src/sources/host_metrics.rs
Comment (bruceg) [reply src/sources/host_metrics.rs:699]: Hmm, looking at it further, I'm surprised this works. AFAIK `block_on` depends on an async runtime being running, but the runtime isn't created until later in `fn prepare_from_opts`. I would have expected this to throw an error.

In any case, this kind of issue is why I would tend to want to move the blocking into the caller, particularly for early init functions, to make it obvious that a runtime must be running.

Comment (eeyun) [reply src/sources/host_metrics.rs:699]: I think that's a very good call @bruceg ! I'll push a fix first thing in the am

File: src/sources/host_metrics.rs
Comment (MOZGIII) [reply src/sources/host_metrics.rs]: Let’s not apply additional constraints - it’s already an opt-in - it makes sense to keep it simple to make it more predictable.

File: src/sources/host_metrics.rs
Comment (MOZGIII) [reply src/sources/host_metrics.rs]: The reason it works is it’s not a `tokio` executor, but `futures` - it simply run the future to completion on the current thread (without `spawn` support, but with the ability to manage concurrent execution within one feature).
@eeyun I didn’t expect this to involve async operation, I assumed we’d just assign values from the env syncronously. I still think we don’t need to go beyond that.

File: src/sources/host_metrics.rs
Comment (MOZGIII) [reply src/sources/host_metrics.rs]: Btw, containers are sort of a well known use case for this, but I’ve seen special dev environments with custom linux-based OS where the procfs and sysfs were mounted in non-standard locations. It’s hardly a case for Vector production use, but there might be users with such a setup out there.

File: src/sources/host_metrics.rs
Comment (MOZGIII) [inline src/sources/host_metrics.rs:32]: What if we move this use to the function itself so it doesn’t have to be kept track of separately?

**Comment by MOZGIII** [comment]
I'm running with `PROCFS_ROOT=/qwerty` (invalid path) and getting no errors. Something must be broken.
It's an early report, I'll continue my testing in the meantime.

Testing with the following settings:

```shell
cat <<'EOF' > cfg.toml
[sources.host_metrics]
type = "host_metrics"
scrape_interval_secs = 5

[sinks.prometheus]
type = "prometheus"
inputs = ["host_metrics"]
address = "0.0.0.0:9090"
EOF
```

Without fs passthrough:

```shell
docker run --rm -it -v "$PWD:$PWD" --cpus 2 -p 9090:9090 test --config "$PWD/cfg.toml"
```

With fs passthrough:

```shell
docker run --rm -it -v "$PWD:$PWD" --cpus 2 -p 9090:9090 -v /proc:/host/proc:ro -e PROCFS_ROOT=/host/proc -v /sys:/host/sys:ro -e SYSFS_ROOT=/host/sys test --config "$PWD/cfg.toml"
```

I Will post updates here as I go.

File: src/sources/host_metrics.rs
Comment (MOZGIII) [inline src/sources/host_metrics.rs:704]: Let's add a debug level print here (and below) to notify users that the non-standard path is picked up for use.

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs]: Not sure what to tell you about that. Nothing in the heim functions we added is explicitly async so I was surprised by this myself.

File: src/sources/host_metrics.rs
Comment (eeyun) [reply src/sources/host_metrics.rs]: This can actually be completely removed. It's a redundant inclusion.

File: src/sources/host_metrics.rs
Comment (bruceg) [reply src/sources/host_metrics.rs]: `fn heim::virt::detect` is async, and digging a bit, what it calls involves some async filesystem calls (reads from proc and sys, etc). Since you removed that bit, there is no longer any need for this to be async, and so no `block_on`.

@MOZGIII I'm confused then. Without tokio, what handles things like non-blocking I/O that are invoked by the futures?

**Review by bruceg — APPROVED** [review]
LGTM once CI checks clear

File: src/sources/host_metrics.rs
Comment (MOZGIII) [reply src/sources/host_metrics.rs]: In my local testing, I figured this branch is likely not executing at all, presumably because the `heim::virt::detect()` is broken.

**Comment by MOZGIII** [comment]
Please wait a bit with the merge, I'd like to finish testing.

**Comment by MOZGIII** [comment]
Btw, here's a tip on what to expect from the metrics (i.e. how to compare on an obvious examle):

```shell
$ docker run --rm -i busybox cat /sys/fs/cgroup/memory/memory.limit_in_bytes
9223372036854771712
$ docker run --rm -i -m 512m busybox cat /sys/fs/cgroup/memory/memory.limit_in_bytes
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
536870912
$ docker run --rm -i -m 512m -v /sys:/host/sys busybox cat /host/sys/fs/cgroup/memory/memory.limit_in_bytes
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
9223372036854771712
```

**Comment by eeyun** [comment]
Aside from our testing - this won't be ready to merge until the upstream module re-exports pass CI and we alter our consumption here. Should be momentarily though.

File: src/sources/host_metrics.rs
Comment (MOZGIII) [reply src/sources/host_metrics.rs]: Non-blocking io can still function as long as there's a reactor loop. (I suspect the terminology I'm used to is different from the naming conventions used in `tokio` - so take this with a grain of salt).
Reactor loop is not necessarily a scheduler and doesn't necessarily handle "task" management and lifecycle - it just has to poll the async function when the Waker signals, and park the thread when the future returns pending. Afaik, In the case of async io, the Waker abstraction is enough to handle the progress.

We essentially only need something like tokio if we want to schedule tasks over a thread pool with a scheduler, or if we need a timer (which doesn't have to be handled at the reactor level, but in the current ecosystem it is). IO operations are portable across the "async runtimes".

File: src/sources/host_metrics.rs
Comment (MOZGIII) [inline src/sources/host_metrics.rs:704]: I think there was an overhaul of logs in which we moved things to structured logging form from the formatting form as much as possible. Not sure if it's a strict rule now to use the structured form, cc @jamtur01

```suggestion
                info!(
                    message = "PROCFS_ROOT is set in envvars. Using custom for procfs.",
                    ?procfs_root
                );
```
It would look something like this if we were to used structured form.

The same should probably be applied to the logs below.

**Review by MOZGIII — APPROVED** [review]

**Pull request merged**
